geode-core/src/main/java/org/apache/geode/internal/cache/LocalRegion.java

```diff
@@ -2109,24 +2109,26 @@ int entryCount(Set<Integer> buckets, boolean estimate) {
   @Override
   public int getRegionSize() {
     synchronized (getSizeGuard()) {
-      int result = getRegionMap().size();
-      // if this is a client with no tombstones then we subtract the number
-      // of entries being affected by register-interest refresh
-      if (imageState.isClient() && !getConcurrencyChecksEnabled()) {
-        int destroyedEntriesCount = imageState.getDestroyedEntriesCount();
-        if (result < destroyedEntriesCount) {
-          logger.error("Incorrect region size: mapSize={}, destroyedEntriesCount={}.", result,
-              destroyedEntriesCount);
+      synchronized (tombstoneCount) {
+        int result = getRegionMap().size();
+        // if this is a client with no tombstones then we subtract the number
+        // of entries being affected by register-interest refresh
+        if (imageState.isClient() && !getConcurrencyChecksEnabled()) {
+          int destroyedEntriesCount = imageState.getDestroyedEntriesCount();
+          if (result < destroyedEntriesCount) {
+            logger.error("Incorrect region size: mapSize={}, destroyedEntriesCount={}.", result,
+                destroyedEntriesCount);
+          }
+          return result - destroyedEntriesCount;
```

Comment:
since we know returning a negative number can cause problems and is not meaningful, I think we should return 0  after logging instead of returning a negative number